lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php

```diff
@@ -386,14 +391,10 @@ private function addInheritedRelations(ClassMetadata $subClass, ClassMetadata $p
             }
 
             //$subclassMapping = $mapping;
-            if (! isset($mapping['inherited']) && ! $parentClass->isMappedSuperclass) {
+            if (! isset($mapping['inherited']) && ! $declaredInMappedSuperclass) {
```

Comment:
I should mention that I'm not sure whether this line should be changed, since I'm unfamiliar with the `['inherited']` property.

If this change is correct, I see that there are similar lines at L356 and L405 that should probably be updated in the same way.

Reply:
If I understand correctly, the `['inherited']` value is set when a property is inherited from something other than a MappedSuperclass.

Looking at this scenario:
```
  Root:            Joined/Single Table Inheritance Entity
  Child:            Mapped Superclass
  Grandchild:       Mapped Superclass
  GreatGrandChild:  Entity
```

If I understand correctly, putting a toOne association on `Child` should not set the ['inherited'] key of the association. However, this change does set that key when the inherited associations are added to GreatGrandChild, since `$declaredInMappedSuperclass` will be false. I think therefore that this line should be reverted? Can someone confirm?

This also indicates to me that the variable `$declaredInMappedSuperclass` is misleading, and the code should probably be adjusted to the following:
```php
// L384
$declaredInParent =  && $mapping['declared'] === $parentClass->name;
if ($parentClass->isMappedSuperclass && $declaredInParent) {
    if ($mapping['type'] & ClassMetadata::TO_MANY && ! $mapping['isOwningSide']) {
        throw MappingException::illegalToManyAssociationOnMappedSuperclass($parentClass->name, $field);
    }

    $mapping['sourceEntity'] = $subClass->name;
}

//$subclassMapping = $mapping;
if (! isset($mapping['inherited']) && ! $parentClass->isMappedSuperclass) {
    $mapping['inherited'] = $parentClass->name;
}
```

If my analysis is correct, I can fix the code and add a relevant test to catch this issue in the future.

Reply:
I've pushed the changes mentioned above as commit 4095ff6, including an extra test to cover this case.